### PR TITLE
[Post-LocalSGD] Comm hook that runs allreduce for the first K steps

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52636 [Post-LocalSGD] Comm hook that runs allreduce for the first K steps**

This DDP communication hook runs allreduce to average gradients for the first K steps, and then stops any communication afterwards.

This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage. It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.

A new test `test_allreduce_first_k_steps_ddp_comm_hook_nccl` is added. Also refactors test_c10d.py by removing the `_test_allreduce_ddp_comm_hook_nccl`, which is already included by `_test_stateless_ddp_comm_hooks_nccl`, and renaming `test_default_ddp_comm_hooks_nccl`  as `test_stateless_ddp_comm_hooks_nccl`.

Differential Revision: [D26576654](https://our.internmc.facebook.com/intern/diff/D26576654/)

### CODE REVIEW

File: torch/lib/c10d/reducer.cpp [hunk 1]
```diff
@@ -1259,7 +1259,11 @@ void Reducer::finalize_backward() {
           bucket.future_work,
           "Expected bucket.future_work not to be null. "
           "This may indicate that communication hook was not properly installed.");
-      bucket.future_work->wait();
+      // Normally need to wait until the completion of communication hook,
+      // unless the communication hook skips allreduce and only does local accumulation for (post-)local SGD.
+      if (!bucket.future_work->completed()) {
+        bucket.future_work->wait();
+      }
```
Comment (pritamdamania87) [inline torch/lib/c10d/reducer.cpp:1266]: I'm not sure why we need this change for post local sgd? Just calling `wait()` should work even for post local SGD since we're retuning a completed future?

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 2]
```diff
@@ -65,3 +67,78 @@ def decompress(fut):
...
+
+    # Run allreduce in the first K steps.
+    if state.step < state.k:
+        state.maybe_increase_step(bucket)
+        return _allreduce_fut(group_to_use, input_tensors[0])
+    else:
+        # Afterwards K steps, directly return the input tensors as the output future.
+        ret_fut = torch.futures.Future()
+        ret_fut.set_result(input_tensors)
+        return ret_fut
```
Comment (pritamdamania87) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:144]: Shouldn't we implement local SGD here where we now average gradients every H steps? I'm not sure why we leave this upto the user? It should be easy to implement that logic here right?

File: test/distributed/test_c10d.py [hunk 3]
```diff
@@ -360,16 +361,23 @@ def test_compare_set(self):
...
             client_store.set("new_key", "new_value0")
             client_store.compare_set("new_key", "new_value1", "new_value0")
         except Exception:
-            logging.error('Caught exception: \n{}exiting process with exit code: {}'
-                          .format(traceback.format_exc(), MultiProcessTestCase.TEST_ERROR_EXIT_CODE))
+            logging.error(
+                "Caught exception: \n{}exiting process with exit code: {}".format(
+                    traceback.format_exc(), MultiProcessTestCase.TEST_ERROR_EXIT_CODE
+                )
+            )
```
Comment (pritamdamania87) [inline test/distributed/test_c10d.py:375]: It's usually easier to review PRs if it doesn't contain formatting changes like these. Usually its better to send a separate PR with formatting changes.

File: test/distributed/test_c10d.py [hunk 4]
```diff
@@ -3822,9 +3825,36 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
+    ):
+        """
+        This unit test verifies whether `allreduce_first_k_steps_hook`
+        can give the same result with the case of no hook registered.
+        """
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        # Get GPU model with the hook registered.
+        for use_error_feedback, warm_start in product([True, False], [True, False]):
```
Comment (pritamdamania87) [inline test/distributed/test_c10d.py:3839]: What is the purpose of this for loop? Doesn't look like we're using `use_error_feedback` or `warm_start`?

File: test/distributed/test_c10d.py [hunk 5]
```diff
@@ -3822,9 +3825,36 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
+            )
+            gpu_model = self._gpu_model_with_ddp_comm_hook(
+                process_group,
+                default.allreduce_first_k_steps_hook,
+                gradient_as_bucket_view,
+                state,
+            )
+
+            # check whether the grads are equal to what DDP without hook would return.
+            self._run_and_verify_hook(gpu_model, 8, 0.25 * torch.ones(2, 2))
```
Comment (pritamdamania87) [inline test/distributed/test_c10d.py:3853]: This only runs one iteration, so I'm not sure how this is testing an algorithm that probably needs multiple iterations?  Also, mathematically post local SGD wouldn't be equivalent to DDP without hooks in general right?

File: torch/lib/c10d/reducer.cpp (hunk 1 above)
Comment (SciPioneer) [reply torch/lib/c10d/reducer.cpp:1266]: We cannot call `wait()` again on a completed future returned by post local sgd. The training will be stuck.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:144]: Local SGD needs to average parameters instead of gradients, and parameter averaging should be executed after `optimizer.step()`, so actually it cannot be implemented in the context of DDP at this time.

File: test/distributed/test_c10d.py (hunk 5 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:3853]: Currently `_run_and_verify_hook` only runs one step. To test it more thoroughly, need to provide another util function.

This test only tests one special case at this time, where post local SGD is equivalent to DDP if there are too few steps and local SGD does not start at all.

What about adding a TODO here to add another test later? I prefer exploring some local SGD algorithms first. It's still unclear how it will work with DDP.

File: test/distributed/test_c10d.py (hunk 4 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:3839]: Extra code copied from PowerSGD unit test. Removed now.

File: test/distributed/test_c10d.py (hunk 3 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:375]: Yeah, I somehow messed it up. Let me revert the irrelevant changes.

File: torch/lib/c10d/reducer.cpp (hunk 1 above)
Comment (SciPioneer) [reply torch/lib/c10d/reducer.cpp:1266]: Per offline discussion, will make `ivalue::Future::wait()` here conditional and avoid extra overheads due to an additional unnecessary lock.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 6]
```diff
@@ -65,3 +67,78 @@ def decompress(fut):
...
+        self.k = k
+        self.step = 0
+
+        logging.info(
+            "Allreduce-first-k-steps hook runs allreduce for the first {} steps.".format(
+                self.k,
+            )
+        )
+
+    def maybe_increase_step(self, bucket):
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:99]: nit: I remember we have a similar function for other hooks that keep track of iteration. Could we consolidate that logic, maybe by a simple counter function/class?

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py [hunk 7]
```diff
@@ -65,3 +67,78 @@ def decompress(fut):
...
+    It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.
+
+    Args:
+        state (AllReduceFirstKStepsState): State information to run allreduce for the first K steps.
+        bucket (dist._GradBucket): Bucket that stores a 1D flattened gradient tensor that batches multiple per-variable tensors.
+            Note that since DDP comm hook only supports single process single device mode at this time,
+            only exactly one tensor is stored in this bucket.
+
+    Returns:
+        Future handler of the communication, which updates the gradients in place.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:125]: nit: mention that this will just a completed future with input tensors (local grads) in the case step > k?

File: test/distributed/test_c10d.py [hunk 8]
```diff
@@ -3819,6 +3799,36 @@ def _test_default_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
+        # if there are too few steps and local SGD does not start at all.
+        # Need to add another test case that involves local SGD.
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        # Get GPU model with the hook registered.
+        state = default.AllReduceFirstKStepsState(
+            process_group=process_group,
+            # Should be equivalent to allreduce for the first K steps.
+            k=1,
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:3820]: I feel like we should test > 1 iterations and with a k > 1. For example, can we test n iterations and k = m, and for the last n - m steps, verify that the gradients are local-only?

File: test/distributed/test_c10d.py [hunk 9]
```diff
@@ -3866,18 +3876,18 @@ def _test_builtin_ddp_comm_hooks_nccl(self, gradient_as_bucket_view=False):
...
-        self._test_default_ddp_comm_hooks_nccl()
+    def test_builtin_ddp_comm_hooks_nccl(self):
+        self._test_builtin_ddp_comm_hooks_nccl()
 
     @requires_nccl()
     @skip_if_lt_x_gpu(2)
-    def test_builtin_ddp_comm_hooks_nccl(self):
-        self._test_builtin_ddp_comm_hooks_nccl()
+    def test_allreduce_first_k_steps_ddp_comm_hook_nccl(self):
+        self._test_allreduce_first_k_steps_ddp_comm_hook_nccl()
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:3890]: Should we test with gradient_as_bucket_view=True?

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 6 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:99]: They are similar, applied to different state classes. The main difference is the log info and the variable name that represents the last iteration.

To consolidate them, need to expose a helper function as a new public API. However, this hook is very experimental, as it is still unclear to support post-local SGD with DDP, prefer no public API changes based on this hook at this time.

Added a TODO here.

File: torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py (hunk 7 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:125]: Good catch!

File: test/distributed/test_c10d.py (hunk 8 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:3820]: Agree. Currently `_run_and_verify_hook` only runs one step. To test it more thoroughly, probably need to provide another util function.

Already added a TODO here to add more test cases later. I prefer exploring some local SGD algorithms first. It's still unclear how it will work with DDP. We can make this hook as a private method at this time.

File: test/distributed/test_c10d.py (hunk 9 above)
Comment (SciPioneer) [reply test/distributed/test_c10d.py:3890]: Added

**Comment by zhaojuanmao** [comment]
DDP allows users to run some iteration with context manager "ddp_model.no_sync()", which should be simple enough to run allreduce for the first K steps. 
```
    for i in range(n): 
        if i < k: 
           context = model.no_sync()
       esle: 
           context = contextlib.nullcontext()
      with context: 
           training logic
```

I'm wondering why we need this hook? if users use this hook, it will increase memory usage as well as increase overhead of copy data btw future and bucket?

File: test/distributed/test_c10d.py (hunk 8 above)
Comment (pritamdamania87) [reply test/distributed/test_c10d.py:3833]: I feel we should add the unit tests in the PR where we introduce the feature itself. You can always patch this PR locally and explore algorithms on top of it.

Or if we're only exploring this idea at this time and might not land this change, probably we can tag this PR as [WIP] or something indicating its not ready for review yet.

**Comment by SciPioneer** [comment]
> DDP allows users to run some iteration with context manager "ddp_model.no_sync()", which should be simple enough to run allreduce for the first K steps.
> 
> ```
>     for i in range(n): 
>         if i < k: 
>            context = model.no_sync()
>        esle: 
>            context = contextlib.nullcontext()
>       with context: 
>            training logic
> ```
> 
> I'm wondering why we need this hook? if users use this hook, it will increase memory usage as well as increase overhead of copy data btw future and bucket?

The key difference is that if `no_sync` context manager is used, `optimizer.step()` must be separate out of the training loop. However local SGD needs to include it in the loop.

This hook does not clone any separate tensors. Since the underlying storage of the input tensors will be shared within the hook, I don't think it will increase memory usage.

File: torch/lib/c10d/reducer.cpp (hunk 1 above)
Comment (SciPioneer) [reply torch/lib/c10d/reducer.cpp]: Actually this change will fail some tests in `distributed_test.py`. With the help of @rohan-varma, realized that the returned type is `CUDAFuture`, which will be always a completed future and will still need to call `wait()` for synchronization.

Reverted this change.

**Comment by SciPioneer** [comment]
To submit this PR, still need to write a unit test case where no more gradient averaging after K steps.

Close this PR since I find another better approach that does not need this communication hook.
